pallets/stateful-storage/src/lib.rs

```diff
@@ -430,55 +428,96 @@ impl<T: Config> Pallet<T> {
 		let items: Vec<ItemizedStorageResponse> = page
 			.parse_as_itemized(false)
 			.map_err(|_| Error::<T>::CorruptedState)?
 			.items
 			.iter()
 			.map(|(key, v)| ItemizedStorageResponse::new(*key, v.to_vec()))
 			.collect();
-		Ok(ItemizedStoragePageResponse::new(msa_id, schema_id, content_hash, items))
+		Ok(ItemizedStoragePageResponse::new(msa_id, schema_id, page.get_hash(), items))
 	}
 
-	fn check_schema_and_grants(
-		provider_key: T::AccountId,
-		state_owner_msa_id: MessageSourceId,
+	pub fn check_payload_expiration(
+		current_block: T::BlockNumber,
+		payload_expire_block: T::BlockNumber,
+	) -> Result<(), DispatchError> {
+		ensure!(payload_expire_block > current_block, Error::<T>::ProofHasExpired);
+		let max_supported_signature_block = Self::mortality_block_limit(current_block);
+		ensure!(payload_expire_block < max_supported_signature_block, Error::<T>::ProofNotYetValid);
+		Ok(())
+	}
+
+	/// Verify the `signature` was signed by `signer` on `payload` by a wallet
+	/// Note the `wrap_binary_data` follows the Polkadot wallet pattern of wrapping with `<Byte>` tags.
+	///
+	/// # Errors
+	/// * [`Error::InvalidSignature`]
+	///
+	pub fn check_signature(
+		signature: &MultiSignature,
+		signer: &T::AccountId,
+		payload: Vec<u8>,
+	) -> DispatchResult {
+		let key = T::ConvertIntoAccountId32::convert((*signer).clone());
+		let wrapped_payload = wrap_binary_data(payload);
+
+		ensure!(signature.verify(&wrapped_payload[..], &key), Error::<T>::InvalidSignature);
+
+		Ok(())
+	}
+
+	/// The furthest in the future a mortality_block value is allowed
+	/// to be for current_block
+	/// This is calculated to be past the risk of a replay attack
+	fn mortality_block_limit(current_block: T::BlockNumber) -> T::BlockNumber {
+		current_block + T::BlockNumber::from(T::MortalityWindowSize::get())
+	}
+
+	fn check_schema(
 		schema_id: SchemaId,
-		payload_location: PayloadLocation,
+		expected_payload_location: PayloadLocation,
+		is_payload_signed: bool,
+		is_deleting: bool,
 	) -> DispatchResult {
-		let provider_msa_id = T::MsaInfoProvider::ensure_valid_msa_key(&provider_key)
-			.map_err(|_| Error::<T>::InvalidMessageSourceAccount)?;
 		let schema =
 			T::SchemaProvider::get_schema_by_id(schema_id).ok_or(Error::<T>::InvalidSchemaId)?;
+
+		// Ensure that the schema's payload location matches the expected location.
 		ensure!(
-			schema.payload_location == payload_location,
+			schema.payload_location == expected_payload_location,
 			Error::<T>::SchemaPayloadLocationMismatch
 		);
 
+		// Ensure that the schema allows signed payloads.
+		// If so, calling extrinsic must be of signature type.
+		if schema.settings.contains(&SchemaSetting::SignatureRequired) {
+			ensure!(is_payload_signed, Error::<T>::SchemaNotSupported);
+		}
+
+		// Ensure that the schema does not allow deletion for AppendOnly SchemaSetting.
+		if schema.settings.contains(&SchemaSetting::AppendOnly) {
+			ensure!(!is_deleting, Error::<T>::SchemaNotSupported);
+		}
```

Comment:
@aramikm  what you think, wanted to keep it simple, given we will circle back at validation